src/net/net.zig

```diff
@@ -73,7 +72,140 @@ pub const SocketAddr = union(enum(u8)) {
+                    {
+                        const ov = @mulWithOverflow(result.V6.scope_id, 10);
+                        if (ov[1] != 0) return error.Overflow;
+                        result.V6.scope_id = ov[0];
+                    }
+                    {
+                        const ov = @addWithOverflow(result.V6.scope_id, digit);
+                        if (ov[1] != 0) return error.Overflow;
+                        result.V6.scope_id = ov[0];
+                    }
+                } else {
+                    return error.InvalidCharacter;
+                }
+            } else if (c == ':') {
+                if (!saw_any_digits) {
+                    if (abbrv) return error.InvalidCharacter; // ':::'
+                    if (i != 0) abbrv = true;
+                    @memset(ip_slice[index..], 0);
+                    ip_slice = tail[0..];
+                    index = 0;
+                    continue;
+                }
+                if (index == 14) {
+                    return error.InvalidEnd;
+                }
+                ip_slice[index] = @as(u8, @truncate(x >> 8));
+                index += 1;
+                ip_slice[index] = @as(u8, @truncate(x));
+                index += 1;
+
+                x = 0;
+                saw_any_digits = false;
+            } else if (c == '%') {
+                if (!saw_any_digits) {
+                    return error.InvalidCharacter;
+                }
+                scope_id = true;
+                saw_any_digits = false;
+            } else if (c == '.') {
+                if (!abbrv or ip_slice[0] != 0xff or ip_slice[1] != 0xff) {
+                    // must start with '::ffff:'
+                    return error.InvalidIpv4Mapping;
+                }
+                const start_index = std.mem.lastIndexOfScalar(u8, buf[0..i], ':').? + 1;
+                const addr = (std.net.Ip4Address.parse(buf[start_index..], 0) catch {
+                    return error.InvalidIpv4Mapping;
+                }).sa.addr;
+                ip_slice = result.V6.ip.octets[0..];
+                ip_slice[10] = 0xff;
+                ip_slice[11] = 0xff;
+
+                const ptr = std.mem.sliceAsBytes(@as(*const [1]u32, &addr)[0..]);
+
+                ip_slice[12] = ptr[0];
+                ip_slice[13] = ptr[1];
+                ip_slice[14] = ptr[2];
+                ip_slice[15] = ptr[3];
+                return result;
+            } else {
+                const digit = try std.fmt.charToDigit(c, 16);
+                {
+                    const ov = @mulWithOverflow(x, 16);
+                    if (ov[1] != 0) return error.Overflow;
+                    x = ov[0];
+                }
+                {
+                    const ov = @addWithOverflow(x, digit);
+                    if (ov[1] != 0) return error.Overflow;
+                    x = ov[0];
+                }
+                saw_any_digits = true;
+            }
+        }
 
+        if (!saw_any_digits and !abbrv) {
```

Comment:
more clear if change to 
```
if (!abbrv) { // if abbrv then rest is all zeros 
  // otherwise require the full address 
  if (index != 14 || !saw_any_digits) { return incomplete } 
  // do final parsing here 
} else { 
  // rest of slice is zeros logic here 
}
```